src/validation.cpp

```diff
@@ -2405,6 +2294,195 @@ bool BitcoindRPCCheck(const bool init)
+    // the other validity checks to RPC calls.
+
+    const std::vector<std::vector<unsigned char> >& stack = pegin_witness.stack;
+    // Must include all elements
+    if (stack.size() != 6) {
+        return false;
+    }
+
+    // Get output value. Special 8 byte length to capture possible bitcoin values
+    CAmount value = CScriptNum(stack[0], true, 8).getint64();
+
+    // Get asset type
+    if (stack[1].size() != 32) {
+        return false;
+    }
+    CAsset asset(stack[1]);
+
+    // Get genesis blockhash
+    if (stack[2].size() != 32) {
+        return false;
+    }
+    uint256 gen_hash(stack[2]);
+
+    // Get witness program
+    CScript witness_program(stack[3].begin(), stack[3].end());
+    int version = -1;
+    std::vector<unsigned char> witnessProgram;
+    if (!witness_program.IsWitnessProgram(version, witnessProgram)) {
+        return false;
+    }
+
+    // Get serialized transaction
+    Sidechain::Bitcoin::CTransactionRef pegtx;
+	try {
+		CDataStream pegtx_stream(stack[4], SER_NETWORK, PROTOCOL_VERSION);
+        pegtx_stream >> pegtx;
+        if (!pegtx_stream.empty()) {
+	        return false;
+		}
+	} catch (std::exception& e) {
+        // Invalid encoding of transaction
+        return false;
+    }
+
+    // Get txout proof
+    Sidechain::Bitcoin::CMerkleBlock merkle_block;
+    std::vector<uint256> txHashes;
+    std::vector<unsigned int> txIndices;
+
+	try {
+        CDataStream merkleBlockStream(stack[5], SER_NETWORK, PROTOCOL_VERSION);
+        merkleBlockStream >> merkle_block;
+        if (!merkleBlockStream.empty() || !CheckBitcoinProof(merkle_block.header.GetHash(), merkle_block.header.nBits)) {
+           return false;
+        }
+        if (merkle_block.txn.ExtractMatches(txHashes, txIndices) != merkle_block.header.hashMerkleRoot || txHashes.size() != 1) {
+            return false;
+        }
+	} catch (std::exception& e) {
+        // Invalid encoding of merkle block
+        return false;
+    }
+
+    // Check that transaction matches txid
+    if (pegtx->GetHash() != prevout.hash) {
+        return false;
+    }
+
+    // Check that the merkle proof corresponds to the txid
+    if (prevout.hash != txHashes[0]) {
+        return false;
+    }
+
+    // Check the genesis block corresponds to a valid peg (only one for now)
+    if (gen_hash != Params().ParentGenesisBlockHash()) {
+        return false;
+    }
+
+    // Check the asset type corresponds to a valid pegged asset (only one for now)
+    if (asset != Params().GetConsensus().pegged_asset) {
+        return false;
+    }
+
+    // Check the transaction nout/value matches
+    if (prevout.n >= pegtx->vout.size() || value != pegtx->vout[prevout.n].nValue) {
+        return false;
+    }
+
+    // Check that the witness program matches the p2ch on the transaction output
+    CScript tweaked_fedpegscript = calculate_contract(Params().GetConsensus().fedpegScript, witness_program);
```

Comment:
This is assuming that elements script and bitcoin script are the same, or at least similar enough for this to work. That could change in the near future.